src/store_client.cc

```diff
@@ -499,58 +517,126 @@ store_client::fileRead()
     assert(!flags.disk_io_pending);
     flags.disk_io_pending = true;
 
+    // mem->swap_hdr_sz is zero here during initial read(s)
+    const auto nextStoreReadOffset = NaturalSum<int64_t>(mem->swap_hdr_sz, nextHttpReadOffset()).value();
+
+    // XXX: If fileRead() is called when we do not yet know mem->swap_hdr_sz,
+    // then we must read from disk offset zero to learn it. We cannot compute
+    // correct HTTP response start offset on disk without it. However, late
+    // startSwapin() calls imply that the assertion below may fail.
+    Assure(mem->swap_hdr_sz > 0 || !nextStoreReadOffset);
+
+    // TODO: Remove this assertion. Introduced in 1998 commit 3157c72, it
+    // assumes that swapped out memory is freed unconditionally, but we no
+    // longer do that because trimMemory() path checks lowestMemReaderOffset().
+    // It is also misplaced: We are not swapping out anything here and should
+    // not care about any swapout invariants.
     if (mem->swap_hdr_sz != 0)
         if (entry->swappingOut())
-            assert(mem->swapout.sio->offset() > copyInto.offset + (int64_t)mem->swap_hdr_sz);
+            assert(mem->swapout.sio->offset() > nextStoreReadOffset);
```

Comment:
Do you agree with this TODO analysis and conclusion? This question has a low priority.

Reply:
Agreed. Also, this assertion looks unreachable (at least currently) since we do not swapIn() while swapping out.

Reply:
> we do not swapIn() while swapping out

Even when using UFS-based cache_dirs?

Reply:
You are right, with UFS we can read from disk while writing. I checked that this assertion is reacheable for non-SMP squid with UFS dir.

Reply:
Glad we are on the same page. I hope we will remove this assertion (addressing the added TODO) as a followup cleanup project. If others insist, we can also remove it in this branch, but it feels like a somewhat risky out-of-scope change that deserves a dedicated commit.